README.md
# KOSMO-POLY6-PICO-DCO 
This repository contains KICAD Schematics, PCB and front panel to build a 6 voice Raspberry Pi Pico Based Digitaly Controlled Oscilator for the "Polykit" Inspired 6 Channel Synth in Kosmo Format [Look Mum No Computer LMNK]. The credit for the desigin and code goes to the original Project by Jan Knipper AKA PolyKit [PolyKit DCO](https://github.com/polykit/pico-dco)

This project requires a Raspbery Pi Pico MCU.

## About KOSMO-POLY6
The [POLY6](https://github.com/twinturbo/KOSMO-POLY6) GIT contains the details of the porject as a whole, this is one module.

Other modules in the line up are

- KOSMO-POLY6-PICO-DCO - Complete [ DCO ](https://github.com/twinturbo/KOSMO-POLY6-PICO-DCO)
- KOSMO-POLY6-DRUID-DCO -(TBD).
- KOSMO-POLY6-JACKS - Complete. [ JACKS ](https://github.com/twinturbo/KOSMO-POLY6-JACKS)
- KOSMO-POLY6-MIX - In design.
- KOSMO-POLY6-ADSR - In Production with JLCPCB.
- KOSMO-POLY6-VCA - In design.
- KOSMO-POLY6-CONTROLS - (TBD).
- KOSMO-POLY6 Filter - (TBD).
- KOSMO-POLY6-PG Programmer - (TBD)
- KOSMO-POLY6-MetricST-Front-Panel - (TBD)

# KOSMO-POLY6-PICO-DCO Details

The DCO utilises a Raspbery PI Pico using it's PWM Pins to generate a highly accurate frequency and gate signals. It can be controlled by USB MIDI or Standard 5 Pin serial MIDI. The analog oscillator part is based on the Juno 106 and generates sawtooth and square wave signal with a 10Vpp amplitude. Amplitude compensation is done by a smoothed PWM signal coming from the Pico.

## Key features

- Digitally controlled analog oscillator using a Raspberry Pi Pico
- Up to six voices
- Voice stacking             ------ [ requires sub control board ] See > Not Done
- Detuning of voices         ------ [ requires sub control board ] See > Not Done
- Frequency modulation (FM)  ------ [ requires sub control board ] See > Not Done
- Separate gate signal 
- DIN MIDI serial input
- USB MIDI input
- MIDI pitch bend (Via MIDI CC )
- Portamento with adjustable speed ( Via MIDI CC )
- Voltage and Midi telltale LED's
- Power pass through for sub modules.

## How does Jan's Sound & Look
These are examples of how the originator Jan Knipper's sound.
This is how it sounds: 

- [Ramp sample](https://soundcloud.com/polykit/pico-dco-ramp) 
- [Pulse sample](https://soundcloud.com/polykit/pico-dco-pulse) 
- [Polyphonic sample](https://soundcloud.com/polykit/pico-dco-polyphonic)

And in use with Jan 
- Youtube[Tube](https://www.youtube.com/watch?v=vj-DSh6yfM0)
- Youtube[Tube](https://www.youtube.com/watch?v=go1mErOdtsU)

## Schematics and PCB
Schematics in Kicad format can be found for the PCB and Front Panel

[MAIN-BOARD](HARDWARE/MAIN-BOARD) 
[MAIN-BOARD](HARDWARE/FRONT_PANEL) 

Gerber files for the PCB and front panel can be found [here](HARDWARE\PCB-PRINTS).
 

## Pictures 
![KOSMO-POLY6-PICO SCHEMATIC ](/Documentation/schematic-V0.1.1.png)
![KOSMO-POLY6-PICO PCB ](/Documentation/PCB-V0.1.1.png)
![KOSMO-POLY6-PICO Front Panel ](/Documentation/master-fp-V0.1.0.png)
![Kosmo-POLY6-PICO IO](/Documentation/IO-Notes.png)

## Installation on the paspbery Pi Pico (simple)
The software has been written in C++ and compiled into a uf2 format for the PICO, 

Two files are provided in the software/build folder along with the source code in the software folder

### Disabled Options
pico-dco-optdisab.uf2 is built with the options for FM, STACK and Detune disabled. This is ideal if you don't have these inputs connected.

//#define USE_ADC_STACK_VOICES // gpio 28 (adc 2)

//#define USE_ADC_DETUNE       // gpio 27 (adc 1)

//#define USE_ADC_FM           // gpio 26 (adc 0)

### Enabled Options
pico-dco-options.uf2 has the above options enabled.

#define USE_ADC_STACK_VOICES // gpio 28 (adc 2)

#define USE_ADC_DETUNE       // gpio 27 (adc 1)

#define USE_ADC_FM           // gpio 26 (adc 0)



### Writing to the MCU.

You will need the file on your computer, and whilst plugging in the PICO to USB hold the `BOOTSEL` button. 

The computer should recognise it as a Mass storage device, Copy desired UF2 file  onto the USB mass storage device. The PICO will reboot into operating mode. Note that if you run the bootsel again the file will not be there, it's not possible to confirm the upload.

## Usage

After building the module and installing the Pico should register as USB MIDI device. Alternatively serial MIDI input is available. The DCO listens to note on/note off messages on MIDI channel 1. Pitch bend is also supported. Portamento can be enabled by MIDI CC message #65, portamento time can be adjusted by CC message #5.

MIDI Device in Cakewalk
![Kosmo-POLY6-PICO IO](/Documentation/MIDICAKE.png)

The 6 notes play all the time, hence there is a need for an ADSR and VCA to close the note off. Each new note played replaces the oldest in the group in a cyclic fashion.


## Detuning, stacking and FM

The built-in analog-to-digital converter can be used to modulate these parameters. You can add an external control voltage (0-3.3V) to that input pins. These features can be disabled by commenting (`//`) the following lines:

As mentioned above, these are not implemented in hardware on this module but a seperate control module needs to be built. 

#define USE_ADC_STACK_VOICES // gpio 28 (adc 2)

#define USE_ADC_DETUNE       // gpio 27 (adc 1)

#define USE_ADC_FM           // gpio 26 (adc 0)

Hardware wise the FM needs to be biased at 1.675v for zero input , the notes on this can be found in the source code **loop void adc_task()**

## Quirks

Initialy the voices will be floating on startup, which if directly outputed ( not through an ADSR/VCA ) may present as noise, as each of the voices is initialised by a key press the noise will go away. 

## Changes 
TwinTurbo 12/3/23 - 
Ref: > https://lookmumnocomputer.discourse.group/t/raspberry-pi-pico-dco/3421/90
updated source code from 0x90 to 0x9F



# References
The Original 
https://github.com/polykit/pico-dco

https://blog.thea.codes/the-design-of-the-juno-dco/

https://electricdruid.net/roland-juno-dcos/

https://www.youtube.com/watch?v=yYnQYF_Xa8g

https://github.com/raspberrypi/pico-examples/tree/master/pio/pwm

https://github.com/raspberrypi/pico-examples/tree/master/pio/pio_blink

https://github.com/raspberrypi/pico-examples/tree/master/pwm/hello_pwm

https://qiita.com/jamjam/items/f2fdd5c072ff348fd809

https://github.com/infovore/pico-example-midi

https://learn.sparkfun.com/tutorials/midi-tutorial/all

http://www.music-software-development.com/midi-tutorial.html

http://midi.teragonaudio.com/tech/midispec/run.htm


ChangeLog
##########################################
# Change Log KOSMO-POLY6-PICO-DCO        #
################################################################################
# S=Schematic P=PCB F=Front                                                    #
# Num   Date       Time    Who   SFP  Change                                   #
################################################################################
#  #1   00/00/00   12:00   RC    					                           #
#  #2   24/01/23   21:12   RC    S    Added Jacks, Pins, Footprints,           #
#                                     Relaying Components                      #
#  #3   26/01/23   20:23   RC         SPLIT OUT MAIN AND JACK HARDWARE         #
#  #4   26/01/23   21:00   RC    Sm   Main Schematic Tidy                      #
#  #5   26/01/23   21:30   RC    Pm   Main PCB Laid out                        #
#  #6   27/01/23   13:52   RC    Sm   Added grounds , changed labels from sub  #
#  #7   27/01/23   13:52   RC    Sm/pm Added Midi in teltale/shaper            #
#  #8   27/01/23   13:52   RC    SPj  Laid out Jack Schematic                  #
#  #9   27/01/23   16:00   RC    Fm   Front Panel Done - pot name / layer fx   #
#  #10  27/01/23   16:00   RC    Fj   Front Panel V0.1.0 Done bar layer fix    #
#  #11	28/01/23   16:00   RC    SPmj Added power Through                      #
#  #12  28/01/23   16:30   RC    Pm   Added Fills and Text                     #
#  #13  28/01/23   19:33   RC    PFmj All Finished on V0.1.0 To Print          #
#  #14  28/01/23   19:35   RC    PFmj   PRINT V0.1.0                           #
#  #15	16/02/23   08:22   RC	 SMPM V0.1.1 GND on LED's and R4               #
#  #16  16/02/23   08:22   RC    SP R4/PWM Pot values updatged to 5.1K and 10K #
#  #17  16/02/23   20:00   RC    SP U9 74LS14 incorect swapped for 74HC14      #
#  #19  16/02/23   20:20   RC    SP U9 has no 3.3v or GND added man route      #
#  #20  16/02/23   20:40   RC    SP R92 Link not joined to U9-4 "RX"           #
#  #21  17/02/23   08:08   RC    P D9 rotate for better rouring                #
#  #22  06/03/23   21:00   RC    Slpit jack project to seperate repository     #
#  #23  06/03/23   21:00   RC    P increased track gap on audio lines          #   
#  #24  08/03/23   08:09   RC    SP added PG1 pin to go to programmer          #
################################################################################





SOFTWARE/pico-dco.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "pico/stdlib.h"
#include "hardware/gpio.h"
#include "hardware/adc.h"
#include "hardware/pio.h"
#include "hardware/clocks.h"
#include "pico-dco.pio.h"
#include "hardware/pwm.h"
#include "bsp/board.h"
#include "tusb.h"
#include "hardware/uart.h"

#define NUM_VOICES 6
#define MIDI_CHANNEL 1
#define USE_ADC_STACK_VOICES // gpio 28 (adc 2)
#define USE_ADC_DETUNE       // gpio 27 (adc 1)
#define USE_ADC_FM           // gpio 26 (adc 0)

uint STACK_VOICES = 1;
float DETUNE = 0.0f, LAST_DETUNE = 0.0f;
float FM_VALUE = 0.0f;  
float LAST_FM = 0.0f;

// Scale factor for FM. Controls how intense the effect is at maximum input voltage.
// Units: Hertz.
float FM_INTENSITY = 5.0f;          

const float BASE_NOTE = 440.0f;
const uint8_t RESET_PINS[NUM_VOICES] = {13, 8, 12, 9, 11, 10};
const uint8_t RANGE_PINS[NUM_VOICES] = {16, 19, 15, 18, 14, 17};
const uint8_t GATE_PINS[NUM_VOICES] = {2, 3, 4, 5, 6, 7};
const uint8_t VOICE_TO_PIO[NUM_VOICES] = {0, 0, 0, 0, 1, 1};
const uint8_t VOICE_TO_SM[NUM_VOICES] = {0, 1, 2, 3, 0, 1};
const uint16_t DIV_COUNTER = 1250;
uint8_t RANGE_PWM_SLICES[NUM_VOICES];

uint32_t VOICES[NUM_VOICES];
uint8_t VOICE_NOTES[NUM_VOICES];
uint8_t VOICE_GATE[NUM_VOICES];

uint8_t NEXT_VOICE = 0;
uint32_t LED_BLINK_START = 0;
PIO pio[2] = {pio0, pio1};
uint8_t midi_serial_status = 0;
uint16_t midi_pitch_bend = 0x2000, last_midi_pitch_bend = 0x2000;
bool portamento = false;
uint8_t portamento_time = 0, portamento_start = 0, portamento_stop = 0;
float portamento_cur_freq = 0.0f;

void init_sm(PIO pio, uint sm, uint offset, uint pin);
void set_frequency(PIO pio, uint sm, float freq);
float get_freq_from_midi_note(uint8_t note);
void led_blinking_task();
uint8_t get_free_voice();
void usb_midi_task();
void serial_midi_task();
void note_on(uint8_t note, uint8_t velocity);
void note_off(uint8_t note);
void voice_task();
void adc_task();
long map(long x, long in_min, long in_max, long out_min, long out_max);

int main() {
    board_init();
    tusb_init();

    // use more accurate PWM mode for buck-boost converter
    gpio_init(23);
    gpio_set_dir(23, GPIO_OUT);
    gpio_put(23, 1);

    // init serial midi
    uart_init(uart0, 31250);
    uart_set_fifo_enabled(uart0, true);
    gpio_set_function(0, GPIO_FUNC_UART);
    gpio_set_function(1, GPIO_FUNC_UART);

    // pwm init
    for (int i=0; i<NUM_VOICES; i++) {
        gpio_set_function(RANGE_PINS[i], GPIO_FUNC_PWM);
        RANGE_PWM_SLICES[i] = pwm_gpio_to_slice_num(RANGE_PINS[i]);
        pwm_set_wrap(RANGE_PWM_SLICES[i], DIV_COUNTER);
        pwm_set_enabled(RANGE_PWM_SLICES[i], true);
    }

    // pio init
    uint offset[2];
    offset[0] = pio_add_program(pio[0], &frequency_program);
    offset[1] = pio_add_program(pio[1], &frequency_program);
    for (int i=0; i<NUM_VOICES; i++) {
        init_sm(pio[VOICE_TO_PIO[i]], VOICE_TO_SM[i], offset[VOICE_TO_PIO[i]], RESET_PINS[i]);
    }

    // gate gpio init
    for (int i=0; i<NUM_VOICES; i++) {
        gpio_init(GATE_PINS[i]);
        gpio_set_dir(GATE_PINS[i], GPIO_OUT);
    }

    // adc init
    #if defined(USE_ADC_STACK_VOICES) || defined(USE_ADC_DETUNE) || defined(USE_ADC_FM) 
    adc_init();
    #ifdef USE_ADC_DETUNE
    adc_gpio_init(27);
    #endif
    #ifdef USE_ADC_STACK_VOICES
    adc_gpio_init(28);
    #endif
    #ifdef USE_ADC_FM
    adc_gpio_init(26);
    #endif
    #endif

    // init voices
    for (int i=0; i<NUM_VOICES; i++) {
        VOICES[i] = 0;
        VOICE_GATE[i] = 0;
    }

    while (1) {
        tud_task();
        usb_midi_task();
        serial_midi_task();
        voice_task();
        #if defined(USE_ADC_STACK_VOICES) || defined(USE_ADC_DETUNE) || defined(USE_ADC_FM)
        adc_task();
        #endif
        led_blinking_task();
    }
}

void init_sm(PIO pio, uint sm, uint offset, uint pin) {
    init_sm_pin(pio, sm, offset, pin);
    pio_sm_set_enabled(pio, sm, true);
}

void set_frequency(PIO pio, uint sm, float freq) {
    uint32_t clk_div = clock_get_hz(clk_sys) / 2 / freq;
    if (freq == 0) clk_div = 0;
    pio_sm_put(pio, sm, clk_div);
    pio_sm_exec(pio, sm, pio_encode_pull(false, false));
    pio_sm_exec(pio, sm, pio_encode_out(pio_y, 32));
}

float get_freq_from_midi_note(uint8_t note) {
    if (portamento && portamento_start != 0 && portamento_stop != 0) {
        float freq1 = pow(2, (portamento_start-69)/12.0f) * BASE_NOTE;
        float freq2 = pow(2, (portamento_stop-69)/12.0f) * BASE_NOTE;
        if (portamento_cur_freq == 0) {
            portamento_cur_freq = freq1;
        } else {
            if (freq1 < freq2) {
                portamento_cur_freq += 1.0f/(portamento_time+1);
                if (portamento_cur_freq > freq2) portamento_cur_freq = freq2;
            } else {
                portamento_cur_freq -= 1.0f/(portamento_time+1);
                if (portamento_cur_freq < freq2) portamento_cur_freq = freq2;
            }
        }
        return portamento_cur_freq;
    }

    return pow(2, (note-69)/12.0f) * BASE_NOTE;
}

void usb_midi_task() {
    if (tud_midi_available() < 4) return;

    uint8_t buff[4];

    LED_BLINK_START = board_millis();
    board_led_write(true);

    if (tud_midi_packet_read(buff)) {
        if (buff[1] == (0x90 | (MIDI_CHANNEL-1))) {
            if (buff[3] > 0) {
                note_on(buff[2], buff[3]);
            } else {
                note_off(buff[2]);
            }
        }

        if (buff[1] == (0x80 | (MIDI_CHANNEL-1))) {
            note_off(buff[2]);
        }

        if (buff[1] == (0xE0 | (MIDI_CHANNEL-1))) {
            midi_pitch_bend = buff[2] | (buff[3]<<7);
        }

        if (midi_serial_status == (0xB0 | (MIDI_CHANNEL-1))) {
            if (buff[2] == 5) { // portamento time
                portamento_time = buff[3];
            }
            if (buff[2] == 65) { // portamento on/off
                portamento = buff[3] > 63;
            }
        }
    }
}

void serial_midi_task() {
    if (!uart_is_readable(uart0)) return;

    uint8_t lsb = 0, msb = 0;

    uint8_t data = uart_getc(uart0);
    
    LED_BLINK_START = board_millis();
    board_led_write(true);

    // status
    if (data >= 0xF0 && data <= 0xF7) {
        midi_serial_status = 0;
        return;
    }

    // realtime message
    if (data >= 0xF8 && data <= 0xFF) {
        return;
    }

    if (data >= 0x80 && data <= 0xEF) {
        midi_serial_status = data;
    }
//https://lookmumnocomputer.discourse.group/t/raspberry-pi-pico-dco/3421/90
// updated from 0x90 to 0x9F RC 12/3/23
    if (midi_serial_status >= 0x80 && midi_serial_status <= 0x9F ||
        midi_serial_status >= 0xB0 && midi_serial_status <= 0xBF || // cc messages
        midi_serial_status >= 0xE0 && midi_serial_status <= 0xEF) {

        lsb = uart_getc(uart0);
        msb = uart_getc(uart0);
    }

    if (midi_serial_status == (0x90 | (MIDI_CHANNEL-1))) {
        if (msb > 0) {
            note_on(lsb, msb);
        } else {
            note_off(lsb);
        }
    }

    if (midi_serial_status == (0x80 | (MIDI_CHANNEL-1))) {
        note_off(lsb);
    }

    if (midi_serial_status == (0xE0 | (MIDI_CHANNEL-1))) {
        midi_pitch_bend = lsb | (msb<<7);
    }

    if (midi_serial_status == (0xB0 | (MIDI_CHANNEL-1))) {
        if (lsb == 5) { // portamento time
            portamento_time = msb;
        }
        if (lsb == 65) { // portamento on/off
             portamento = msb > 63;
        }
    }
}

void note_on(uint8_t note, uint8_t velocity) {
    if (STACK_VOICES < 2) {
        for (int i=0; i<NUM_VOICES; i++) {
            if (VOICE_NOTES[i] == note && VOICE_GATE[i]) return; // note already playing
        }
    }
    for (int i=0; i<STACK_VOICES; i++) {
        uint8_t voice_num = get_free_voice();

        VOICES[voice_num] = board_millis();
        VOICE_NOTES[voice_num] = note;
        VOICE_GATE[voice_num] = 1;

        float freq = get_freq_from_midi_note(note) * (1 + (pow(-1, i) * DETUNE));
        set_frequency(pio[VOICE_TO_PIO[voice_num]], VOICE_TO_SM[voice_num], freq);
        // amplitude adjustment
        pwm_set_chan_level(RANGE_PWM_SLICES[voice_num], pwm_gpio_to_channel(RANGE_PINS[voice_num]), (int)(DIV_COUNTER*(freq*0.00025f-1/(100*freq))));
        // gate on
        gpio_put(GATE_PINS[voice_num], 1);
    }
    if (portamento) {
        if (portamento_start == 0) {
            portamento_start = note;
            portamento_cur_freq = 0.0f;
        } else {
            portamento_stop = note;
        }
    }
    last_midi_pitch_bend = 0;
}

void note_off(uint8_t note) {
    // gate off
    for (int i=0; i<NUM_VOICES; i++) {
        if (VOICE_NOTES[i] == note) {
            gpio_put(GATE_PINS[i], 0);

            //VOICE_NOTES[i] = 0;
            VOICES[i] = 0;
            VOICE_GATE[i] = 0;
        }
    }
    if (portamento_stop == note) {
        portamento_start = portamento_stop;
        portamento_stop = 0;
        portamento_cur_freq = 0.0f;
    }
    /*
    if (portamento_start == note) {
        portamento_stop = 0;
        portamento_cur_freq = 0.0f;
    }
    */
}

uint8_t get_free_voice() {
    uint32_t oldest_time = board_millis();
    uint8_t oldest_voice = 0;

    for (int i=0; i<NUM_VOICES; i++) {
        uint8_t n = (NEXT_VOICE+i)%NUM_VOICES;

        if (VOICE_GATE[n] == 0) {
            NEXT_VOICE = (n+1)%NUM_VOICES;
            return n;
        }

        if (VOICES[i]<oldest_time) {
            oldest_time = VOICES[i];
            oldest_voice = i;
        }
    }

    NEXT_VOICE = (oldest_voice+1)%NUM_VOICES;
    return oldest_voice;
}

void voice_task() {
    if (midi_pitch_bend != last_midi_pitch_bend || DETUNE != LAST_DETUNE || portamento || FM_VALUE != LAST_FM) {

        last_midi_pitch_bend = midi_pitch_bend;
        LAST_DETUNE = DETUNE;
        LAST_FM = FM_VALUE;

        for (int i=0; i<NUM_VOICES; i++) {

            float freq = get_freq_from_midi_note(VOICE_NOTES[i]) * (1 + (pow(-1, i) * DETUNE));

            freq += FM_VALUE * FM_INTENSITY; // Add linear frequency modulation
            
            freq = freq-(freq*((0x2000-midi_pitch_bend)/67000.0f));
            set_frequency(pio[VOICE_TO_PIO[i]], VOICE_TO_SM[i], freq);
            pwm_set_chan_level(RANGE_PWM_SLICES[i], pwm_gpio_to_channel(RANGE_PINS[i]), (int)(DIV_COUNTER*(freq*0.00025f-1/(100*freq))));
        }
    }
}

void adc_task() {
    uint16_t raw;

    #ifdef USE_ADC_DETUNE
    adc_select_input(1);
    raw = adc_read();
    DETUNE = map(raw, 0, 4095, 0, 50)/1000.0f;
    #endif

    #ifdef USE_ADC_STACK_VOICES
    adc_select_input(2);
    raw = adc_read();
    STACK_VOICES = map(raw, 0, 4000, 1, NUM_VOICES);
    #endif

    #ifdef USE_ADC_FM

    adc_select_input(0);
    raw = adc_read();

    // This input assumes a centre FM value of 2^11 = 2048, as the 
    // ADCs are unsigned. This can be overcome in hardware with a fixed 
    // voltage offset. The range of the Pico's ADCs is 3.3V, so a fixed 
    // offset of 1.65V is needed.
    int signed_raw = raw - (1 << 11);
    FM_VALUE = (float) signed_raw / (float) (1 << 11);
    #endif

}

long map(long x, long in_min, long in_max, long out_min, long out_max) {
  return (x - in_min) * (out_max - out_min) / (in_max - in_min) + out_min;
}

void led_blinking_task() {
    if (board_millis() - LED_BLINK_START < 50) return;
    board_led_write(false);
}
